include/seahorn/Expr/EvalModel.hh

```diff
@@ -0,0 +1,302 @@
+#pragma once
+
+#include "seahorn/Expr/EvalUtils.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprGmp.hh"
+#include "seahorn/Expr/ExprOpBind.hh"
+#include "seahorn/Expr/ExprOpBinder.hh"
+
+#include <math.h>
+#include <stdlib.h>
+
+namespace expr {
+namespace eval {
+
+/**
+ * Bit vector structure. Holds the number(as type T) and its width
+ *
+ * \note bools are stored as 1/0 with a width of 1
+ */
+template <typename T> struct BvNum {
+  T num = 0;
+  unsigned width = 0;
+
+  /// \brief empty constructor
+  BvNum() : num(0), width(0) {}
+
+  /// \brief bool constructor.
+  BvNum(bool a) : num(a), width(1) {}
+
+  BvNum(T numArg, unsigned widthArg) : num(numArg), width(widthArg) {}
+
+  bool operator<(const BvNum<T> &bnum) const { return num < bnum.num; }
+
+  bool operator==(const BvNum<T> &bnum) const {
+    return (num == bnum.num) && (width == bnum.width);
+  }
+};
+
+/**
+ * Holds the data for an array.
+ *
+ * In EvalModel, the array expressions are mapped to this struct
+ *
+ */
+template <typename T> struct Arr {
+  bool m_hasDefaultVal = false;
+  BvNum<T> m_defaultVal;
+
+  unsigned m_valWidth = 0;
+
+  std::map<BvNum<T>, BvNum<T>> m_map; ///< index, value
+
+  Arr(BvNum<T> defaultVal)
+      : m_hasDefaultVal(true), m_defaultVal(defaultVal),
+        m_valWidth(defaultVal.width) {}
+  Arr(unsigned valWidth) : m_hasDefaultVal(false), m_valWidth(valWidth) {}
+};
+
+/**
+ * Holds the data for a lambda. Maps bound vars to their value
+ *
+ * The purpose of this is to keep bound vars within the scope of the lambda. The
+ * same bound var expression could have different values in different lambdas
+ *
+ * In EvalModel, lambda expressions are mapped to this struct
+ *
+ */
+template <typename T> struct Lambda {
+  std::map<Expr, BvNum<T>> m_map; ///< bound var, value
+};
+
+/**
+ * \class EvalModel . The base class for the eval models
+ *
+ * \details The purpose is to generate and store values for constants, bound
+ * variables, and arrays. All 3 types have their own get/set methods
+ *
+ * \note the get methods generate a value if one does not already exist
+ *
+ *  To create a new model, extend EvalModel, and override generateNum(). See
+ *  EvalModel_Rand for an example
+ */
+template <typename Type> class EvalModel {
+
+protected:
+  /// maps the constants/bvars to their values
+  std::map<Expr, BvNum<Type>> m_cache;
```

Comment:
is this `cache` or `data`? Cache is usually something that can be re-computed and might become stale.
I think this is the actual data that underpins the model